bittensor/metagraph.py

```diff
@@ -567,112 +553,365 @@ async def _poll_pubkey(self, pubkey):
-        r""" Returns true if the passed key and vals are set on chain.
+        except SubscribeUnknownError as e:
+            logger.error('Subscription through an unknown error: {}', e)
+            return False
+        
+        except SubscribeTimeout as e:
+            logger.error('Subscription timeout {}', e)
+            return False
+
+        except Exception as e:
+            logger.error('Subscription threw uncaught error {}', e)
+            return False
+        
+    async def _try_async_subscribe(self, timeout: int):
+        r""" Makes subscription attempts to the chain, continuing to attempt until timeout and finally waiting for inclusion.
+
+        Args:
+            timeout (int):
+                Time to wait before subscription times out.
+
+        Raises:
+            SubscribeSuccess (Exception):
+                Raised when the subscription is a success before the timeout
+
+            SubscribeUnknownError (Exception):
+                UnknownError during subscription.
+
+            SubscribeTimeout (Exception):
+                Raised when the attempted subscription fails after timeout.
         """
-        cmap = {}
-        chain_keys = await self.subtensor_client.weight_keys(self.__keypair.public_key)
-        chain_vals = await self.subtensor_client.weight_vals(self.__keypair.public_key)
-        for key, val in list(zip(chain_keys, chain_vals)):
-            cmap[key] = val
-        for key, val in list(zip(keys, vals)):
-            if key not in cmap:
-                return False
-            if cmap[key] != val:
-                return False 
-        return True
+        start_time = time.time()
+        # ---- Make Subscription transaction ----
+        while True:
+            try:
+                await self.subtensor_client.subscribe(self._config.axon.remote_ip, self._config.axon.port)
+                break
+
+            except Exception as e:
+                if (time.time() - start_time) > timeout:
+                    # --- Timeout during emit call ----
+                    message = "Timed-out with Unknown Error while trying to make the subscription call. With last exception {}".format(e)
+                    raise SubscribeUnknownError(message)
+
+                else:
+                    # --- Wait for inclusion, no error.
+                    logger.trace('Error while attempting subscription {}', e)
+                    continue
+
+        # ---- Wait for inclusion ----
+        while True:
+            try:
+                # ---- Request info from chain ----
+                info = await self.subtensor_client.neurons(self.__keypair.public_key)
+
+            except Exception as e:
+                # ---- Catch errors in request ----
+                raise SubscribeUnknownError("Subscription threw unknown exception {}".format(e))
+
+            if info != None:
+                # ---- Subscription was a success ----
+                raise SubscribeSuccess("Subscription success")
+
```

Comment:
Not sure why we're raising an exception if subscription was a success...?

Reply:
basically they act as dispatch results not as true errors